project/vision_backend/tasks.py

```diff
@@ -323,6 +330,83 @@ def _handle_job_result(job_res: JobReturnMsg):
         logger.info("Collected job: {} with pk: {}".format(task_name, pk))
 
 
+def train_fail(job_return_msg: JobReturnMsg):
+    job_token = job_return_msg.original_job.tasks[0].job_token
+    pk = th.decode_spacer_job_token(job_token)[0]
+
+    try:
+        classifier = Classifier.objects.get(pk=pk)
+    except Classifier.DoesNotExist:
+        # This should be an edge case, where the user reset the classifiers or
+        # deleted the source before training could complete.
+        logger.info(
+            "Training failed for classifier {}, although the classifier"
+            " was already deleted.".format(job_token))
+        return
+
+    classifier.status = Classifier.TRAIN_ERROR
+    classifier.save()
+    logger.info(
+        "Training failed for classifier {}.".format(job_token))
+
+    # Sometimes training fails because features were temporarily de-synced
+    # with the points. We'll take different actions depending on the number of
+    # train attempts that failed in a row.
+    source = classifier.source
+    last_classifiers = source.classifier_set.order_by('-pk')[:5]
+    train_fails_in_a_row = 0
+    for clf in last_classifiers:
+        if clf.status == Classifier.TRAIN_ERROR:
+            train_fails_in_a_row += 1
+        else:
+            break
+
+    if train_fails_in_a_row <= 1:
+        # Only 1 fail so far.
+        # Sometimes the features are getting re-extracted, and just didn't
+        # finish yet. We'll hope for that here.
+        submit_classifier.apply_async(
+            args=[source.pk],
+            eta=now() + timedelta(hours=3))
+    elif 2 <= train_fails_in_a_row <= 3:
+        # Go through all the source's images and force-extract features.
+        for image in source.image_set.all():
+            reset_features.apply_async(
```

Comment:
You could also call `submit_features` directly with the `Force` flag set. https://github.com/beijbom/coralnet/blob/master/project/vision_backend/tasks.py#L35

Reply:
Nevermind -- I like your way better.